xdem/coreg/affine.py

```diff
@@ -54,147 +54,728 @@
+    """
 
-        :returns: Estimated y-values with the same shape as the given x-values
+    def _calculate_slope_and_aspect_nuthkaab(dem: NDArrayf) -> tuple[NDArrayf, NDArrayf]:
         """
-        return parameters[0] * np.cos(parameters[1] - x_values) + parameters[2]
+        Calculate the tangent of slope and aspect of a DEM, in radians, as needed for the Nuth & Kaab algorithm.
+
+        For now, this method using the gradient is more efficient than slope/aspect derived in the terrain module.
 
-    def residuals(parameters: tuple[float, float, float], y_values: NDArrayf, x_values: NDArrayf) -> NDArrayf:
+        :param dem: A numpy array of elevation values.
+
+        :returns:  The tangent of slope and aspect (in radians) of the DEM.
         """
-        Get the residuals between the estimated and measured values using the given parameters.
 
-        err(x, y) = est_y(x) - y
+        # Gradient implementation
+        # # Calculate the gradient of the slope
+        gradient_y, gradient_x = np.gradient(dem)
+        slope_tan = np.sqrt(gradient_x**2 + gradient_y**2)
+        aspect = np.arctan2(-gradient_x, gradient_y)
+        aspect += np.pi
 
-        :param parameters: The a, b, and c parameters to use for the estimation.
-        :param y_values: The measured y-values.
-        :param x_values: The measured x-values
+        # Terrain module implementation
+        # slope, aspect = xdem.terrain.get_terrain_attribute(
+        #     dem, attribute=["slope", "aspect"], resolution=1, degrees=False
+        # )
+        # slope_tan = np.tan(slope)
+        # aspect = (aspect + np.pi) % (2 * np.pi)
 
-        :returns: An array of residuals with the same shape as the input arrays.
-        """
-        err = estimate_ys(x_values, parameters) - y_values
-        return err
+        return slope_tan, aspect
+
+    # If inputs are both point clouds, raise an error
+    if isinstance(ref_elev, gpd.GeoDataFrame) and isinstance(tba_elev, gpd.GeoDataFrame):
+
+        raise TypeError(
+            "The Nuth and Kääb (2011) coregistration does not support two point clouds, one elevation "
+            "dataset in the pair must be a DEM."
+        )
+
+    # If inputs are both rasters, derive terrain attributes from ref and get 2D dh interpolator
+    elif isinstance(ref_elev, np.ndarray) and isinstance(tba_elev, np.ndarray):
+
+        # Derive slope and aspect from the reference as default
+        slope_tan, aspect = _calculate_slope_and_aspect_nuthkaab(ref_elev)
+
+    # If inputs are one raster and one point cloud, derive terrain attribute from raster and get 1D dh interpolator
+    else:
+
+        if isinstance(ref_elev, gpd.GeoDataFrame):
+            rst_elev = tba_elev
+        else:
+            rst_elev = ref_elev
+
+        # Derive slope and aspect from the raster dataset
+        slope_tan, aspect = _calculate_slope_and_aspect_nuthkaab(rst_elev)
+
+    return slope_tan, aspect
+
+
+def _nuth_kaab_iteration_step(
```

Comment:
Would be good to briefly describe the input arguments (in case one needs to do some debugging in the future :wink:)

Reply:
Added! I leave out the ones copied from `fit()` or `apply()` to avoid redundancy, but forgot these new ones!